pkg/descheduler/controllers/migration/arbitrator/arbitrator.go

```diff
@@ -0,0 +1,294 @@
+
+	"github.com/koordinator-sh/koordinator/apis/scheduling/v1alpha1"
+	"github.com/koordinator-sh/koordinator/pkg/descheduler/apis/config"
+	"github.com/koordinator-sh/koordinator/pkg/descheduler/framework"
+)
+
+const (
+	AnnotationPassedArbitration = "descheduler.koordinator.sh/passed-arbitration"
+)
+
+var enqueueLog = klog.Background().WithName("eventHandler").WithName("arbitratorImpl")
+
+type arbitratorImpl struct {
+	waitingCollection map[types.UID]*v1alpha1.PodMigrationJob
+	workQueue         workqueue.RateLimitingInterface
+	interval          time.Duration
+
+	sorts                 []SortFn
+	nonRetryablePodFilter framework.FilterFunc
+	retryablePodFilter    framework.FilterFunc
+
+	client        client.Client
+	eventRecorder events.EventRecorder
+	mu            sync.Mutex
+}
+
+// Add adds a PodMigrationJob to the waitingCollection of arbitratorImpl.
+// It is safe to be called concurrently by multiple goroutines.
+func (a *arbitratorImpl) Add(job *v1alpha1.PodMigrationJob) {
+	a.mu.Lock()
+	defer a.mu.Unlock()
+	if a.waitingCollection == nil {
+		klog.Errorf("waitingCollection is nil")
+	}
+	a.waitingCollection[job.UID] = job.DeepCopy()
+}
+
+// Arbitrate starts the goroutine to arbitrate jobs periodically.
+func (a *arbitratorImpl) Arbitrate(stopCh <-chan struct{}) {
+	klog.Infof("Start Arbitrator Arbitrate Goroutine")
+	for {
+		a.doOnceArbitrate()
+		select {
+		case <-stopCh:
+			return
+		case <-time.After(a.interval):
+		}
+	}
+}
+
+func (a *arbitratorImpl) WithSortFn(sort SortFn) {
+	a.sorts = append(a.sorts, sort)
+}
+
+// sort stably sorts jobs, outputs the sorted results and corresponding ranking map.
+func (a *arbitratorImpl) sort(jobs []*v1alpha1.PodMigrationJob, podOfJob map[*v1alpha1.PodMigrationJob]*corev1.Pod) []*v1alpha1.PodMigrationJob {
+	for _, sortFn := range a.sorts {
+		jobs = sortFn(jobs, podOfJob)
+	}
+	return jobs
+}
+
+// filter calls nonRetryablePodFilter and retryablePodFilter to filter operations on PodMigrationJobs.
+func (a *arbitratorImpl) filter(jobs []*v1alpha1.PodMigrationJob, podOfJob map[*v1alpha1.PodMigrationJob]*corev1.Pod) {
+	for _, job := range jobs {
+		pod := podOfJob[job]
+		if pod != nil {
+			if a.nonRetryablePodFilter != nil && !a.nonRetryablePodFilter(pod) {
+				a.updateFailedJob(job, pod)
+				continue
+			}
+			if a.retryablePodFilter != nil && !a.retryablePodFilter(pod) {
+				continue
+			}
+		}
+		a.updatePassedJob(job)
+	}
+}
+
+// updatePassedJob does something after PodMigrationJob passed the filter.
+func (a *arbitratorImpl) updatePassedJob(job *v1alpha1.PodMigrationJob) {
+	// add annotation AnnotationPassedArbitration
+	if job.Annotations == nil {
+		job.Annotations = map[string]string{}
+	}
+	job.Annotations[AnnotationPassedArbitration] = "true"
+	err := a.client.Update(context.TODO(), job)
+	if err != nil {
+		klog.Errorf("failed to update job %v, err: %v", fmt.Sprintf("%v/%v", job.Namespace, job.Name), err)
+	}
+
+	if a.workQueue != nil {
```

Comment:
If updated the job to api server, I think it's unnecessary to operate the workqueue. Am I right?

Reply:
I just update the pmj for adding `AnnotationPassedArbitration` to the passed pmjs. But it may trigger another `update event`. If we remove the workQueue, it may look a little strange for all of the create events will be transferred to update event.